File: bindings/csharp/WickraCompile/Compiler.cs
using System.Runtime.InteropServices;
using System.Text;
using System.Text.Json;

namespace Wickra.Compile;

/// <summary>
/// Compiles a Wickra strategy spec into a standalone deployable and a
/// deterministic manifest, driven by JSON commands over the Wickra C ABI.
/// Construct one, drive it with command JSON (<c>compile</c>, <c>targets</c>,
/// <c>version</c>, <c>artifact_bytes</c>, <c>reset</c>) and read back the
/// response JSON — the same protocol as the CLI and every other binding.
/// </summary>
public sealed class Compiler : IDisposable
{
    private readonly CompilerHandle _handle;

    /// <summary>Construct a compiler handle.</summary>
    public Compiler()
    {
        IntPtr ptr = Native.wickra_compile_new();
        if (ptr == IntPtr.Zero)
        {
            throw new InvalidOperationException("wickra-compile: allocation failed");
        }
        _handle = new CompilerHandle(ptr);
    }

    /// <summary>Apply a command JSON and return the response JSON.</summary>
    /// <remarks>
    /// Uses the C ABI's length-out protocol: a first call learns the length, then
    /// the response is read into a caller-owned buffer. Domain errors (a bad
    /// command, an unknown command, an invalid spec) come back in-band as
    /// <c>{"ok":false,...}</c> JSON, not as an exception.
    /// </remarks>
    /// <exception cref="InvalidOperationException">A required argument was unusable or a panic was caught.</exception>
    public string Command(string cmdJson)
    {
        ObjectDisposedException.ThrowIf(_handle.IsInvalid, this);

        byte[] cmd = Utf8(cmdJson);
        IntPtr h = _handle.DangerousGetHandle();
        int n = Native.wickra_compile_command(h, cmd, null, 0);
        if (n < 0)
        {
            throw new InvalidOperationException($"wickra-compile: command failed (code {n})");
        }
        var buf = new byte[n + 1];
        Native.wickra_compile_command(h, cmd, buf, (nuint)buf.Length);
        return Encoding.UTF8.GetString(buf, 0, n);
    }

    /// <summary>Read the raw bytes of a file through the C ABI byte reader.</summary>
    /// <exception cref="InvalidOperationException">The file could not be read.</exception>
    public byte[] ArtifactBytes(string path)
    {
        string cmd = JsonSerializer.Serialize(new Dictionary<string, string>
        {
            ["cmd"] = "artifact_bytes",
            ["path"] = path,
        });
        using JsonDocument doc = JsonDocument.Parse(Command(cmd));
        JsonElement root = doc.RootElement;
        if (root.TryGetProperty("error", out JsonElement err))
        {
            throw new InvalidOperationException($"wickra-compile: {err.GetString()}");
        }
        ulong byteHandle = root.GetProperty("handle").GetUInt64();
        int len = root.GetProperty("len").GetInt32();
        if (len == 0)
        {
            return [];
        }
        var buf = new byte[len];
        IntPtr h = _handle.DangerousGetHandle();
        long n = Native.wickra_compile_artifact_read(h, byteHandle, buf, (nuint)buf.Length);
        if (n < 0)
        {
            throw new InvalidOperationException($"wickra-compile: artifact read failed (code {n})");
        }
        return buf;
    }

    /// <summary>The library version.</summary>
    public static string Version() =>
        Marshal.PtrToStringUTF8(Native.wickra_compile_version()) ?? string.Empty;

    /// <summary>Free the native compiler handle.</summary>
    public void Dispose() => _handle.Dispose();

    /// <summary>Encode a string as NUL-terminated UTF-8 for the C ABI.</summary>
    private static byte[] Utf8(string s)
    {
        int len = Encoding.UTF8.GetByteCount(s);
        var buf = new byte[len + 1];
        Encoding.UTF8.GetBytes(s, 0, s.Length, buf, 0);
        return buf;
    }
}

/// <summary>A safe handle owning a native compiler pointer.</summary>
internal sealed class CompilerHandle : SafeHandle
{
    public CompilerHandle(IntPtr handle)
        : base(IntPtr.Zero, ownsHandle: true) => SetHandle(handle);

    public override bool IsInvalid => handle == IntPtr.Zero;

    protected override bool ReleaseHandle()
    {
        Native.wickra_compile_free(handle);
        return true;
    }
}

File: bindings/java/src/main/java/org/wickra/compile/Compiler.java
package org.wickra.compile;

import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.ValueLayout;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Compiles a Wickra strategy spec into a standalone deployable and a
 * deterministic manifest, driven by JSON commands over the Wickra C ABI
 * (FFM/Panama). Construct one, drive it with command JSON ({@code compile},
 * {@code targets}, {@code version}, {@code artifact_bytes}, {@code reset}) and
 * read back the response JSON — the same protocol as the CLI and every other
 * binding.
 */
public final class Compiler implements AutoCloseable {
    private static final Pattern HANDLE = Pattern.compile("\"handle\"\\s*:\\s*(\\d+)");
    private static final Pattern LEN = Pattern.compile("\"len\"\\s*:\\s*(\\d+)");

    private MemorySegment handle;

    /** Construct a compiler handle. */
    public Compiler() {
        try {
            MemorySegment created = (MemorySegment) Native.NEW.invokeExact();
            if (created.address() == 0) {
                throw new IllegalStateException("wickra-compile: allocation failed");
            }
            this.handle = created;
        } catch (RuntimeException | Error e) {
            throw e;
        } catch (Throwable t) {
            throw new RuntimeException(t);
        }
    }

    /**
     * Apply a command JSON and return the response JSON. Uses the C ABI's
     * length-out protocol: a first call learns the length, then the response is
     * read into a caller-owned buffer. Domain errors (a bad command, an unknown
     * command, an invalid spec) come back in-band as {@code {"ok":false,...}}
     * JSON, not as an exception.
     */
    public String command(String cmdJson) {
        if (handle == null) {
            throw new IllegalStateException("compiler is closed");
        }
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment cmd = arena.allocateFrom(cmdJson);
            int len = (int) Native.COMMAND.invokeExact(handle, cmd, MemorySegment.NULL, 0L);
            if (len < 0) {
                throw new IllegalStateException("wickra-compile: command failed (code " + len + ")");
            }
            MemorySegment buf = arena.allocate(len + 1L);
            int written = (int) Native.COMMAND.invokeExact(handle, cmd, buf, (long) (len + 1));
            return buf.getString(0);
        } catch (RuntimeException | Error e) {
            throw e;
        } catch (Throwable t) {
            throw new RuntimeException(t);
        }
    }

    /**
     * Read the raw bytes of a file through the C ABI byte reader.
     *
     * @throws IllegalStateException if the file could not be read
     */
    public byte[] artifactBytes(String path) {
        if (handle == null) {
            throw new IllegalStateException("compiler is closed");
        }
        String response = command("{\"cmd\":\"artifact_bytes\",\"path\":\"" + escape(path) + "\"}");
        if (response.contains("\"ok\":false") || response.contains("\"error\"")) {
            throw new IllegalStateException("wickra-compile: " + response);
        }
        long byteHandle = extract(HANDLE, response, "handle");
        int len = (int) extract(LEN, response, "len");
        if (len == 0) {
            return new byte[0];
        }
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment buf = arena.allocate(len);
            long n = (long) Native.ARTIFACT_READ.invokeExact(handle, byteHandle, buf, (long) len);
            if (n < 0) {
                throw new IllegalStateException("wickra-compile: artifact read failed (code " + n + ")");
            }
            return buf.asSlice(0, n).toArray(ValueLayout.JAVA_BYTE);
        } catch (RuntimeException | Error e) {
            throw e;
        } catch (Throwable t) {
            throw new RuntimeException(t);
        }
    }

    /** The library version. */
    public static String version() {
        try {
            MemorySegment ptr = (MemorySegment) Native.VERSION.invokeExact();
            return ptr.reinterpret(Long.MAX_VALUE).getString(0);
        } catch (Throwable t) {
            throw new RuntimeException(t);
        }
    }

    /** Free the native compiler handle. */
    @Override
    public void close() {
        if (handle != null) {
            try {
                Native.FREE.invokeExact(handle);
            } catch (Throwable t) {
                throw new RuntimeException(t);
            }
            handle = null;
        }
    }

    private static long extract(Pattern pattern, String json, String field) {
        Matcher matcher = pattern.matcher(json);
        if (!matcher.find()) {
            throw new IllegalStateException("wickra-compile: missing " + field + " in response: " + json);
        }
        return Long.parseLong(matcher.group(1));
    }

    /** Escape a string for embedding in a JSON string literal. */
    private static String escape(String s) {
        return s.replace("\\", "\\\\").replace("\"", "\\\"");
    }
}
